### ISSUE TITLE
Potentially unintentional breaking change in 4.0 with Symbol in argument matchers?

### ISSUE THREAD

**Issue by jarthod** [wspurgin/rspec-sidekiq#203]
I noticed a breaking change in 4.0 which does not seem to be listed in https://github.com/wspurgin/rspec-sidekiq/blob/main/CHANGES.md. And so I am wondering if it is intentional or not.

It's about the `have_enqueued_sidekiq_job` matchers, when used with symbols for argument matching. It used to work but now doesn't any more (expects strings). Here is an example spec:
```ruby
expect(WebhookWorker).to have_enqueued_sidekiq_job("http://host.fr", [{
  event: 'check.down',
  time: "2023-03-23T00:00:00Z"
}])
```

And the error message given by 4.0 (passing with 3.1)
```
 expected to have an enqueued WebhookWorker job
   with arguments:
     -["http://host.fr", [{:event=>"check.down", :time=>"2023-03-23T00:00:00Z"}]]
 but have enqueued only jobs
  -JID:75961eff3be0808eeeaf26f9 with arguments:
     -["http://host.fr", [{"event"=>"check.down", "time"=>"2023-03-23T00:00:00Z"}]]
```

If I change all Symbols to Strings it passes in 4.0. That's not a a big deal for me I know Sidekiq uses String anyway, I was mostly using Symbols here for the slightly better hash readability (IMO). But it's a couple hundreds specs to update so before doing that I wanted to make sure this is a wanted and desirable change (in which case it should be added to the CHANGES).

Thanks!

**Comment by wspurgin** [comment]
Thanks for raising it up @jarthod - it wasn't documented but it's the result of using Sidekiq's searlizied arguments (which searlizes to JSON as you noted) in combination with RSpec's `ArgumentListMatcher`. Which let us use the builtin rspec matchers like `hash_including`

I _think_ we could update the base matcher argument "soft" jsonifying to deep stringify keys for any Hash literals detected...

**Comment by wspurgin** [comment]
Alright, so there's a slight complication. For basic Hash literals it'll work fine, _but_ when combining a Hash literal with a builtin, it will force us to mismatch:
```ruby
worker.perform_async({"something" => 1})
expect(worker).to have_enqueued_sidekiq_job({"something" => kind_of(Integer) })
#=> actual soft jsonified arguments [{"something"=>"#<RSpec::Mocks::ArgumentMatchers::KindOf:0x0000000110ce8d88>"}]

#=> Fails but is misleading because the error message formats off the raw expected args
#        expected to have an enqueued Workerk5imtt36ns job
#         with arguments:
#           -[{"something"=>kind of Integer}]
#      but have enqueued only jobs
#         -JID:49bb3dd13193899b0f564aab with arguments:
#           -[{"something"=>1}]
```

So while hash's were actually dumped to JSON and then reloaded previously, continuing to do that in `rspec-sidekiq` will essentially produce those misleading mismatches.

I'm tempted to just say that we should "do less" when it comes to arg manipulation and avoid doing too much magic that obfuscates what args are being expected. And just say this is new "breaking" behavior in 4.0... still if there's a lot of support for this, I think the _most_ I'd be willing to do is add a `with_indifferent_access` when detecting a Hash literal... it would support symbol keys, but symbol values would still mismatch

**Comment by wspurgin** [comment]
@jarthod It's this bit of inherited code from v3 https://github.com/wspurgin/rspec-sidekiq/blob/main/lib/rspec/sidekiq/matchers/base.rb#L240-L251

Which serializes a Hash to JSON then deserializes it to get a "matching" Hash to the Sidekiq Payload. Works fine for a Hash literal with only literal types, but doesn't work when combining RSpec matchers. Calling `with_indifferent_access` there will at least allow matching with symbol keys, but it won't transform values.

**Comment by jarthod** [comment]
Mm I just checked the code and I'm even more confused now:
1. First I don't see where this JSON conversion was placed in 3.1 (https://github.com/wspurgin/rspec-sidekiq/compare/v3.1.0...main) ?
2. This code piece doesn't even work and never calls the JSON conversion because it uses `args.class` in the `case` statement, comparing the class with another class doesn't work with threequals, it must be the instance:
```ruby
> Array === Array
=> false
> Array === []
=> true
```
3. Also the argument is amost always an Array due to the multiple job arguments being matched, changing the code to this (removing `.class` and adding `Array`) seems to "repair" the previous behavior:
```ruby
@jsonified_expected_arguments ||= begin
  expected_arguments.map do |arg|
    case arg
    when Symbol then arg.to_s
    when Hash, Array then JSON.parse(arg.to_json)
    else arg
    end
  end
end
```

Obivously this breaks the rspec matchers yes, but then why is this broken JSON conversion code in the new version if the goal is not to use it as per the comment? I was already using `hash_including('subject', 'issuer', 'from', 'to')` in my `have_enqueued_sidekiq_job` matchers previously and it was working fine so wasn't the previous [`normalize_arguments`](https://github.com/wspurgin/rspec-sidekiq/compare/v3.1.0...main#diff-9b82c7f947515518c9862b4e4dabc31af47c7fc498177e5d9d07fb273eafdc7cL167-L179) method actually smarter?

As said earlier I am fine with updating all my specs to use strings, but I'm still wondering what is the target here and why did this argument normalization needed to change? If I have a better understanding of the goal of this change and the problems encountered I may be able to help too.

**Comment by wspurgin** [comment]
Sorry, I'm still learning all the nooks and crannies of this repo.

I though that `jsonified_expected_arguments` was the original method. But that was from an in-between potential maintainer's commits that I pulled into my first PR.

I'm fine with restoring the old `normalize_arguments` which is really only concerned about stringifying symbols.

**Comment by wspurgin** [comment]
@jarthod if you wanna try #205 out with your codebase and confirm you see what you expect with respect to symbols, I'll merge it in and cut 4.0.1

**Comment by jarthod** [comment]
> Sorry, I'm still learning all the nooks and crannies of this repo

No problem of course, and thanks for taking over the maintenance by the way!

I just tried #205 and it worked as expected indeed, no breaking change with my codebase on this one :tada:

**Comment by jarthod** [comment]
Thanks for the fix! So it's gonna be in `4.0.1` when released, right?

**Comment by wspurgin** [comment]
Yes @jarthod [`4.0.1`](https://rubygems.org/gems/rspec-sidekiq/versions/4.0.1) just released. I appreciate you digging into this with me and for the bug report! Cheers 🍻

**Comment by jarthod** [comment]
Thanks! Cheers :beers:

**Comment by n-rodriguez** [comment]
Hi there! It's broken when using `ActiveJob`:

```ruby
      context "when expected arguments include symbols" do
        let(:active_job) { create_active_job }
        let(:worker_args) { [:foo, {bar: :baz}] }
        it "returns true" do
          active_job.perform_later(*worker_args)
          expect(Sidekiq::Worker).to have_enqueued_sidekiq_job(*worker_args)
        end
      end
```

```sh

**Comment by n-rodriguez** [comment]
@jarthod do you use `ActiveJob` or `Sidekiq::Job` like in https://github.com/sidekiq/sidekiq/wiki/Getting-Started ?

**Comment by jarthod** [comment]
@n-rodriguez indeed but was it working in 3.1 ? Because this breaking change in 4.0.0 was unintentional. So if it also accidentaly fixed some ActiveJob Symbol support that wasn't there previously it's nice, but if we add it back it's still a breaking change for sidekiq users (which is the core of this gem) so can only be "reverted" carefully in a major release ideally because it would impact much more users.

I use `Sidekiq::Job` in my case, with some occasional use of ActiveJob in ActionMailer but when I do I use ActiveJob spec matchers like [have_enqueued_job](https://www.rubydoc.info/gems/rspec-rails/RSpec/Rails/Matchers#have_enqueued_job-instance_method). So I don't test ActiveJob workers with `rspec-sidekiq`.

As said earlier it's fine for me if we choose the breaking change which respects arguments type and doesn't do any conversion, as long as it's indicated as such in the changelog with some migration instructions (namely replacing all symbols in Sidekiq job specs with Strings).
